solr/solrj/src/java/org/apache/solr/client/solrj/impl/Http2SolrClient.java

```diff
@@ -417,67 +419,101 @@ public void send(OutStream outStream, SolrRequest<?> req, String collection) thr
+    }
+    MDCCopyHelper mdcCopyHelper = new MDCCopyHelper();
+    CompletableFuture<NamedList<Object>> future = new CompletableFuture<>();
+    final MakeRequestReturnValue mrrv;
+    final String url;
     try {
-      String url = getRequestPath(solrRequest, collection);
-      InputStreamResponseListener listener =
-          new InputStreamReleaseTrackingResponseListener() {
-            @Override
-            public void onHeaders(Response response) {
-              super.onHeaders(response);
-              executor.execute(
-                  () -> {
-                    InputStream is = getInputStream();
-                    try {
-                      NamedList<Object> body =
-                          processErrorsAndResponse(solrRequest, response, is, url);
-                      mdcCopyHelper.onBegin(null);
-                      log.debug("response processing success");
-                      asyncListener.onSuccess(body);
-                    } catch (RemoteSolrException e) {
-                      if (SolrException.getRootCause(e) != CANCELLED_EXCEPTION) {
+      url = getRequestPath(solrRequest, collection);
+      mrrv = makeRequest(solrRequest, url, true);
+    } catch (SolrServerException | IOException e) {
+      future.completeExceptionally(e);
+      return future;
+    }
+    final ResponseParser parser =
+        solrRequest.getResponseParser() == null ? this.parser : solrRequest.getResponseParser();
+    mrrv.request
+        .onRequestQueued(asyncTracker.queuedListener)
+        .onComplete(asyncTracker.completeListener)
+        .send(
+            new InputStreamResponseListener() {
+              @Override
+              public void onHeaders(Response response) {
+                super.onHeaders(response);
+                InputStreamResponseListener listener = this;
+                executor.execute(
+                    () -> {
+                      InputStream is = listener.getInputStream();
+
+                      // TODO: Original PR had this, but we do not close the stream.
+                      // The legacy implementation did not track the input stream,
+                      // or close it.
+                      // assert ObjectReleaseTracker.track(is);
+
+                      try {
+                        NamedList<Object> body =
+                            processErrorsAndResponse(solrRequest, response, is, url);
+                        mdcCopyHelper.onBegin(null);
+                        log.debug("response processing success");
+                        future.complete(body);
+                      } catch (RemoteSolrException | SolrServerException e) {
                         mdcCopyHelper.onBegin(null);
                         log.debug("response processing failed", e);
-                        asyncListener.onFailure(e);
+                        future.completeExceptionally(e);
+                      } finally {
+                        log.debug("response processing completed");
```

Comment:
for consistency with 2 cases right above, these 2 lines should be reversed.

Reply:
so true...